File: sdks/java/src/main/java/com/backlex/Subscription.java
package com.backlex;

import java.io.InputStream;

/**
 * Handle for an active realtime subscription. {@link #close()} unsubscribes — the
 * same contract as the TS SDK's returned unsubscribe function. Implements
 * {@link AutoCloseable} for try-with-resources.
 */
public final class Subscription implements AutoCloseable {

    private volatile boolean stopped = false;
    private volatile InputStream current;
    private Thread thread;

    void attachThread(Thread t) {
        this.thread = t;
    }

    void setStream(InputStream s) {
        this.current = s;
    }

    boolean isStopped() {
        return stopped;
    }

    @Override
    public void close() {
        stopped = true;
        InputStream s = current;
        if (s != null) {
            try {
                s.close();
            } catch (Exception ignored) {
                // best-effort: closing the stream breaks the blocking read
            }
        }
        if (thread != null) {
            thread.interrupt();
        }
    }
}

File: sdks/kotlin/src/main/kotlin/com/backlex/Subscription.kt
package com.backlex

import java.io.Closeable
import java.io.InputStream

/**
 * Handle for an active realtime subscription. [close] unsubscribes — the same
 * contract as the TS SDK's returned unsubscribe function. Implements [Closeable]
 * for `use { }`.
 */
class Subscription : Closeable {
    @Volatile
    var stopped: Boolean = false
        private set

    @Volatile
    private var current: InputStream? = null
    private var thread: Thread? = null

    internal fun attachThread(t: Thread) {
        thread = t
    }

    internal fun setStream(s: InputStream) {
        current = s
    }

    override fun close() {
        stopped = true
        try {
            current?.close() // closing the stream breaks the blocking read
        } catch (_: Exception) {
        }
        thread?.interrupt()
    }
}
